source/agora/node/Registry.d

```diff
@@ -604,47 +606,37 @@ private struct TypedPayload
     /// UTXO
     public Hash utxo;
 
-    /// Timestamp for expiration
-    public uint expires;
-
     /***************************************************************************
 
-        Make an instance of an `TypedPayload` from DNS payload
-        (Resource record)
+        Make an instance of an `TypedPayload` from URI records
 
         Params:
             rr = DNS Resource record
             pubKeyParser = Delegate for parsing public key from RR domain
 
     ***************************************************************************/
 
-    public static TypedPayload make (ResourceRecord rr)
+    public static TypedPayload make (ResourceRecord[] rrs)
     {
-        auto public_key = rr.name.extractPublicKey();
+        ensure(rrs !is null || rrs.length > 0,
+            "No records are provided to make TypedPayload");
```

Comment:
Redundant condition (null / length).
Also it looks like something the caller should check, not the callee. In the presence of array / nullable data, this is the approach I usually take:

Functions like this can have one of two behavior: They should either be no-op in the face of `null` data, or not no-op.

If they are not no-op, it means `null` is considered an in-band value. An example would be a function that either can do something with the `null` value (rare) or has some sort of fallback (like the ill-fated https://github.com/bosagora/agora/pull/2981).

Some other functions want to be no-ops in the face of `null`. An example of this is `startPeriodicNameRegistration`, where we just `return` and don't do anything if `config.node.registry_address` is not configured.

Now what is this function doing: It is neither no-op (since it throws) nor is the `null` value in-band.
My instinct here is to ask the question: "Should we have reached **this point** or should we have stopped before?". By checking the caller, I see that the check is also performed in the caller, so we should never have reached this in the first place. And IMO, considering the tradeoffs, that's a sound thing to do. Hence, this shouldn't be an `ensure`, it should be an `assert`, because passing out-of-band data should have been caught earlier.

Now this is not an absolute rule: Some functions will accept out-of-band data and then check if they are in-band. But that is usually their sole focus. One good example is Vibe.d, which needs to accept arbitrary string and turn them into a structured view of the HTTP request. If you were to get an `HTTPClientRequest` from Vibe.d with data that violate the HTTP specs, you would be right to call it a Vibe.d bug.